components/kyma-environment-broker/internal/environmentscleanup/service.go

```diff
@@ -69,78 +71,112 @@ func (s *Service) Run() error {
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) getStaleRuntimesByShoots(labelSelector string) ([]runtime, []unstructured.Unstructured, error) {
 	opts := v1.ListOptions{
 		LabelSelector: labelSelector,
 	}
 	shootList, err := s.gardenerService.List(context.Background(), opts)
 	if err != nil {
-		return []unstructured.Unstructured{}, fmt.Errorf("while listing Gardener shoots: %w", err)
+		return []runtime{}, []unstructured.Unstructured{}, fmt.Errorf("while listing Gardener shoots: %w", err)
 	}
 
+	var runtimes []runtime
 	var shoots []unstructured.Unstructured
 	for _, shoot := range shootList.Items {
 		shootCreationTimestamp := shoot.GetCreationTimestamp()
 		shootAge := time.Since(shootCreationTimestamp.Time)
 
-		if shootAge.Hours() >= s.MaxShootAge.Hours() {
-			log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.GetName(), s.MaxShootAge.Hours(), shootAge.Hours())
+		if shootAge.Hours() < s.MaxShootAge.Hours() {
+			continue
+		}
+
+		log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.GetName(), s.MaxShootAge.Hours(), shootAge.Hours())
+		staleRuntime, err := s.shootToRuntime(shoot)
+		if err != nil {
+			s.logger.Info("found a shoot without accounts references")
 			shoots = append(shoots, shoot)
+			continue
 		}
+
+		runtimes = append(runtimes, *staleRuntime)
 	}
 
-	return shoots, nil
+	return runtimes, shoots, nil
 }
 
-func (s *Service) getRuntimes(shoots []unstructured.Unstructured) []runtime {
-	var runtimes []runtime
-	for _, st := range shoots {
-		shoot := gardener.Shoot{st}
-		runtimeID, ok := shoot.GetAnnotations()[shootAnnotationRuntimeId]
-		if !ok {
-			s.logger.Error(fmt.Errorf("shoot %q has no runtime-id annotation", shoot.GetName()))
-			continue
-		}
-
-		accountID, ok := shoot.GetLabels()[shootLabelAccountId]
-		if !ok {
-			s.logger.Error(fmt.Errorf("shoot %q has no account label", shoot.GetName()))
-			continue
-		}
+func (s *Service) shootToRuntime(st unstructured.Unstructured) (*runtime, error) {
+	shoot := gardener.Shoot{st}
+	runtimeID, ok := shoot.GetAnnotations()[shootAnnotationRuntimeId]
+	if !ok {
+		return nil, fmt.Errorf("shoot %q has no runtime-id annotation", shoot.GetName())
+	}
 
-		runtimes = append(runtimes, runtime{
-			ID:        runtimeID,
-			AccountID: accountID,
-		})
+	accountID, ok := shoot.GetLabels()[shootLabelAccountId]
+	if !ok {
+		return nil, fmt.Errorf("shoot %q has no account label", shoot.GetName())
 	}
 
-	return runtimes
+	return &runtime{
+		ID:        runtimeID,
+		AccountID: accountID,
+	}, nil
 }
 
 func (s *Service) cleanUp(runtimesToDelete []runtime) error {
-	kebInstancesToDelete, err := s.getInstancesForRuntimes(runtimesToDelete)
-	if err != nil {
-		err = fmt.Errorf("while getting instance IDs for Runtimes: %w", err)
-		s.logger.Error(err)
-		return err
+	kebInstancesToDelete := make([]internal.Instance, 0)
+
+	for _, rt := range runtimesToDelete {
+		instance, err := s.getInstancesForRuntimes([]runtime{rt})
```

Comment:
Why we use this function in so limited way selecting instances one by one instead of selection the set.

Reply:
Implementation get array of runtime ids and selects as follows:
```
		Select("*").
		From(InstancesTableName).
		Where("runtime_id IN ?", runtimeIdList).
		LoadOne(&instances)
```
so if not all runtime ids are found we should not encounter error but get shorter list.

Reply:
This is the main reason for the change. If only one runtime does not have an instance then an error is returned and the execution fail.

Reply:
Corrected to check the NOT FOUND error explicitly.